lib/init.php

```diff
@@ -57,3 +57,121 @@ function gutenberg_menu() {
 	);
 }
 add_action( 'admin_menu', 'gutenberg_menu', 9 );
+
+if ( ! function_exists( 'html_contains_block' ) ) {
+	/**
+	 * Returns whether the given HTML contains a block
+	 * of the given type and, if provided,
+	 * a given attribute and attribute value.
+	 * 
+	 * Note that it's not possible to search for an attribute
+	 * whose value is `null`.
+	 *
+	 * @param  string  $html The html to search in.
+	 * @param  string  $block_name      Find this block type,
+	 *                                  with an optional "core/" namespace,
+	 *                                  e.g. "paragraph", "core/paragraph",
+	 *                                  "my_plugin/my_block".
+	 * @param  string  $attribute_name  If provided, the block must also
+	 *                                  contain this attribute.
+	 * @param  string  $attribute_value If provided, the given attribute's
+	 *                                  value must also match this.
+	 *
+	 * @return bool                     True if block is found, false otherwise
+	 */
+	function html_contains_block( $html, $block_name, $attribute_name = null, $attribute_value = null ) {
+		$at = 0;
+
+		/**
+		 * This is the same regex as the one used in the block parser.
+		 * It is better to use this solution to look for a block's existence
+		 * in a document compared to having to parsing the blocks in the
+		 * document, avoiding all the performance drawbacks of achieving
+		 * a full representation of block content just to check if one block
+		 * is there.
+		 *
+		 * @see WP_Block_Parser.
+		 */
+		$pattern = sprintf(
+			'~<!--\s+?wp:%s\s+(?P<attrs>{(?:(?:[^}]+|}+(?=})|(?!}\s+/?-->).)*+)?}\s+)?/?-->~s',
+			preg_quote( str_replace( 'core/', '', $block_name ), '~' )
+		);
```

Comment:
you probably already know this, but we need to expand this a bit because then we'd overlook posts containing `core/template`, for example. although the editor prunes the `core/` namespace, it's perfectly valid to have it written in the HTML.

given the fact that we're only looking for a single block type I recommend we hard-code the regex pattern to match the sought block name. we can do that by breaking apart the block name into its parts, different than the code I shared above about joining them.

```php
// check my string-index math
$solidus_at       = strpos( $block_name, '/' );
$sought_namespace = false === $solidus_at ? 'core' : substr( $block_name, 0, $solidus_at );
$sought_name      = false === $solidus_at ? $block_name : substr( $block_name, $solidus_at + 1 );

$pattern = sprintf(
	'~<!--\s+?wp:((?:%s/)?%s\s+...~',
	preg_quote( $sought_namespace, '~' ),
	preg_quote( $sought_name, '~' )
);
```